File: HW9/HW--2013-11-13--PG9/WordPlay.py
"""
 * WordPlay
 * Nate Mara
 * 2013-11-06
 *
 * Follows a given set of criteria for checking different things
 * about user inputted strings.
 * 
"""
from random import randint

def main():
    # keeps prompting the user for a word until 'Q' is entered

    prompt = "Enter a word for me to analyze (or Q to quit): "
    print

    userInput = raw_input(prompt)

    while userInput != "Q":
        showStats(userInput)
        userInput = raw_input(prompt)

def showStats(word):
    # after the user enters a word, this shows all the info about it

    word = word.upper()
    print
    print "Results for " + word + ":"
    print "Number of vowels: " + str(vowelCount(word))
    if vowelCount(word) != 0 :
        firstVowel = word[indexOfFirstVowel(word)]
        firstVowelPos = indexOfFirstVowel(word)

        print "First vowel is: " + firstVowel + " in position " + str(firstVowelPos)
    
    print "Pig Latin: " + pigLatinize(word)
    print "Ends swapped: " + swapEnds(word)
    print "Reversed: " + reverse(word)
    print "Ends swapped and reversed: " + reverse(swapEnds(word))
    print "Scrambled: " + multiScramble(word)
    printAllSubstrings(word);

def vowelCount(word):
    # returns number of vowels in word using iteration
    numVowels = 0

    for i in word:
        if vowelCheck(i):
            numVowels += 1
    return numVowels

def indexOfFirstVowel(word): 
    # returns index of first vowel in word using iteration
    for i in range(0, len(word)):
        currentChar = word[i]
        
        if vowelCheck(currentChar):
            return i
    return -1

def pigLatinize(word):
    # returns word translated to 'pig latin'

    if indexOfFirstVowel(word) == -1:
        return word + "AY"

    elif indexOfFirstVowel(word) == 0:
        return word + "HAY"

    else:
        untilVowel = word[:indexOfFirstVowel(word)]
        vowelToEnd = word[indexOfFirstVowel(word):]
        return vowelToEnd + untilVowel + "AY"
    
def swapEnds(word):
    # returns the word with the first and last chars switched
    firstChar = word[0]
    lastChar = word[len(word) - 1]

    if len(word) == 1:
        return word

    middle = word[1:len(word) - 1]

    return lastChar + middle + firstChar

def reverse(word): 
    # returns the reverse of the word
    reversedString = ""

    for i in range(0, len(word)):
        index = len(word) - i
        reversedString += word[index - 1]

    return reversedString

def scramble(word): 
    # returns scrambled version of word
    scrambleString = ""
    lenWord = len(word)

    for i in range(0, lenWord):
        myRandInt = randint(0, len(word) - 1)
        randChar = word[myRandInt]
        scrambleString += randChar
        word = word[:myRandInt] + word[myRandInt + 1:]  

    return scrambleString

def vowelCheck(testChar): 
    # returns True if char is vowel, returns False otherwise
    isA = testChar == 'A'
    isE = testChar == 'E'
    isI = testChar == 'I'
    isO = testChar == 'O'
    isU = testChar == 'U'

    if isA | isE | isI | isO | isU:
        return True

    return False

def multiScramble(word): 
    # returns 5 scrambled versions of 'word' each separated by single space
    stringSum = ""
    for i in range(0, 5):
        stringSum += scramble(word) + " "

    return stringSum

def printAllSubstrings(word): 
    # prints all substrings of 'word' with asterisk borders on each side
    # and on top/bottom
    lenWord = len(word)
    count = 0

    print "All substrings of " + word + ":"
    print nStars(lenWord)

    for length in range(1, lenWord + 1):
        for start in range(0, lenWord - length + 1):
            printSubstring(word[start:start + length], lenWord)
            count += 1

    print nStars(lenWord)
    print str(count) + " substrings found."

def nStars(n): 
    # returns the top or bottom border for a word 'n' characters long
    n += 4 
    return "*" * n

def printSubstring(substring, totalLength): 
    # prints a substring with star and space on either side. totalLength refers
    # to the total length of the original word.
    difference = totalLength - len(substring)
    endingSpace = " " * difference

    print "* " + substring + endingSpace + " *"

if __name__ == '__main__':main()

File: HW9/WordPlay.java
/*
 * WordPlay
 * Nate Mara
 * 2013-11-06
 *
 * Follows a given set of criteria for checking different things
 * about user inputted strings.
 * 
 */

import java.util.Scanner;
import java.lang.Math;

public class WordPlay{

	// keeps prompting the user for a word until 'Q' is entered
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);

		String prompt = "Enter a word for me to analyze (or Q to quit): ";
		System.out.println();
		System.out.print(prompt);

		String userInput;
		userInput = in.next();		

		while (!userInput.equals("Q")) {
			showStats(userInput);
			System.out.print(prompt);
			userInput = in.next();
		}

	}
	
	// after the user enters a word, this shows all the info about it
	public static void showStats(String word) {
		word = word.toUpperCase();
		
		System.out.println();
		System.out.println("Results for " + word + ":");
		System.out.println("Number of vowels: " + vowelCount(word));
		if (vowelCount(word) != 0) {
			char firstVowel = word.charAt(indexOfFirstVowel(word));
			int firstVowelPos = indexOfFirstVowel(word);

			System.out.println("First vowel is: " + firstVowel + " in position " + firstVowelPos);
		}

		System.out.println("Pig Latin: " + pigLatinize(word));
		System.out.println("Ends swapped: " + swapEnds(word));
		System.out.println("Reversed: " + reverse(word));
		System.out.println("Ends swapped and reversed: " + reverse(swapEnds(word)));
		System.out.println("Scrambled: " + multiScramble(word));
		printAllSubstrings(word);
	}

	// returns number of vowels in word using iteration
	public static int vowelCount(String word) {
		int numVowels = 0;

		for (int i = 0; i < word.length(); i++) {
			char currentChar = word.charAt(i);
			
			if (vowelCheck(currentChar)) {
				numVowels ++;
			}

		}

		return numVowels;

	}

	// returns index of first vowel in word using iteration
	public static int indexOfFirstVowel(String word) {
		for (int i = 0; i < word.length(); i++) {
			char currentChar = word.charAt(i);
			
			if (vowelCheck(currentChar)) {
				return i;
			}
		}
		return -1;


	}

	// returns word translated to 'pig latin'
	public static String pigLatinize(String word) {
		String untilVowel;
		String vowelToEnd;

		if (indexOfFirstVowel(word) == -1)
			return word + "AY";

		else if (indexOfFirstVowel(word) == 0)
			return word + "HAY";

		else
			untilVowel = word.substring(0, indexOfFirstVowel(word));
			vowelToEnd = word.substring(indexOfFirstVowel(word));
			return vowelToEnd + untilVowel + "AY";
	}
	
	// returns the word with the first and last chars switched
	public static String swapEnds(String word) {
		char firstChar = word.charAt(0);
		char lastChar = word.charAt(word.length() - 1);

		if (word.length() == 1)
			return word;

		String middle = word.substring(1, word.length() - 1);

		return lastChar + middle + firstChar;
	}

	// returns the reverse of the word
	public static String reverse(String word) {
		String reversedString = "";

		for (int i = word.length(); i > 0; i--) {
			reversedString += word.charAt(i - 1);
		}

		return reversedString;
	}

	// returns scrambled version of word
	public static String scramble(String word) {
		String scrambleString = "";
		int randInt;
		char randChar;
		int lenWord = word.length();

		for (int i = 0; i < lenWord; i++) {
			randInt = randomNum(0, word.length());
			randChar = word.charAt(randInt);
			scrambleString += randChar;
			word = word.substring(0, randInt) + word.substring(randInt + 1);	
		}

		return scrambleString;
	}

	// returns true if char is vowel, returns false otherwise
	public static boolean vowelCheck(char testChar) {
		boolean isA, isE, isI, isO, isU;
		isA = testChar == 'A';
		isE = testChar == 'E';
		isI = testChar == 'I';
		isO = testChar == 'O';
		isU = testChar == 'U';

		if (isA || isE || isI || isO || isU) {
			return true;
		}
		return false;
	}
	
	// returns random int in range [low, high) (inclusive low, exclusive high)
	public static int randomNum(int low, int high) {
		double randDouble = Math.random();
		int randInt = (int)(randDouble * high);
		return randInt;
	}

	// returns 5 scrambled versions of 'word' each separated by single space
	public static String multiScramble(String word) {
		String stringSum = "";
		for (int i = 0; i < 5; i++) {
			stringSum += scramble(word) + " ";
		}
		return stringSum;
	}

	// prints all substrings of 'word' with asterisk borders on each side
	// and on top/bottom
	public static void printAllSubstrings(String word) {
		int lenWord = word.length();
		int count = 0;

		System.out.println("All substrings of " + word + ":");
		System.out.println(nStars(lenWord));

		for (int length = 1; length <= lenWord; length++) {
			for (int start = 0; start <= lenWord - length; start++) {
				printSubstring(word.substring(start, start + length), lenWord);
				count ++;
			}
		}

		System.out.println(nStars(lenWord));
		System.out.println(count + " substrings found.");
	}

	// returns the top or bottom border for a word 'n' characters long
	public static String nStars(int n) {
		String starString = "";
		n += 4; 
		while (n > 0) {
			starString += "*";
			n--;
		}
		return starString;
	}

	// prints a substring with star and space on either side. totalLength refers
	// to the total length of the original word.
	public static void printSubstring(String substring, int totalLength) {
		int difference = totalLength - substring.length();
		String endingSpace = "";

		for (int i = 0; i < difference; i++){
			endingSpace += " ";
		}
		System.out.println("* " + substring + endingSpace + " *");
	}

}	
